### ISSUE TITLE
Feat: Enforce custom provider types

### ISSUE THREAD

**Issue by thiagomini** [TrilonIO/eslint-plugin#24]
# Description

We want to allow users to prefer one type of custom provider over the others. For instance, [factory providers over value providers can significantly increase performance](https://github.com/nestjs/nest/pull/12753).

# Acceptance Criteria

Given the options below:

```ts
// ...
rules: {
  '@trilon/enforce-custom-provider-type': [ 
   'warn', {  
     prefer: 'factory'
   }
  ]
}
```

Then, the following code will be considered **Invalid**:

```ts
const customValueProvider: Provider = {
  provide: 'TOKEN',
  useValue: 'some-value' // ⚠️ provider is not of type "factory"
}

const customClassProvider: Provider = {
  provide: AbstractClass,
  useClass: SomeClass  // ⚠️ provider is not of type "factory"
}
```

And the following code is considered **Valid**:

```ts
const factoryProvider: Provider = {
  provide: 'TOKEN',
  useFactory: () => 'some-value'
}
```

**Comment by thiagomini** [comment]
@tuxmachine I'm wondering what we could use to determine that an object is a provider, @tuxmachine 🤔 Maybe we should check for the explicit "Provider" type from Nest? Otherwise, I fear this might get some false-positives from other objects that are not providers
